python/ray/tests/test_runtime_env_conda_and_pip_4.py

```diff
@@ -89,5 +91,24 @@ def f():
         ray.shutdown()
 
 
+@pytest.mark.skipif(
+    "IN_VIRTUALENV" in os.environ or sys.platform != "linux",
+    reason="Requires PR wheels built in CI, so only run on linux CI machines.",
+)
+def test_run_in_virtualenv(cloned_virtualenv):
+    python_exe_path = cloned_virtualenv.python
+    print(python_exe_path)
+
+    # make sure cloned_virtualenv.run will run in virtualenv.
+    cloned_virtualenv.run(
+        f"{python_exe_path} -c 'from ray._private.runtime_env.pip import PipProcessor;"
+        "assert PipProcessor._is_in_virtualenv()'"
+    )
+
+    # Run current test case in virtualenv again.
+    # If the command exit code is not zero, this case will be failed.
+    cloned_virtualenv.run(f"IN_VIRTUALENV=1 python -m pytest {__file__}")
```

Comment:
How does this make `test_run_in_virtualenv` fail?  Do we need to assert that the exit code is nonzero somewhere?

Also, if it fails, will the test output be printed somewhere?  If not, it could be hard to debug.

Reply:
> How does this make `test_run_in_virtualenv` fail? Do we need to assert that the exit code is nonzero somewhere?
> 
> Also, if it fails, will the test output be printed somewhere? If not, it could be hard to debug.

We don't need to `assert exit_code != 0`, this [pytest-plugin will raise an exception when  return_code != 0](https://github.com/man-group/pytest-plugins/blob/75e04eb252f2c025196c0cd8a4292ba29110be41/pytest-shutil/pytest_shutil/workspace.py#L132), and this Exception will construct by this pytest-cmd's stderr, so we may not worry about debug.